File: src/python/xerxes_agent/channels/adapters/bluebubbles.py
# Copyright 2025 The EasyDeL/Xerxes Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License")
"""BlueBubbles iMessage bridge adapter."""

from __future__ import annotations

import typing as tp

from .._helpers import WebhookChannel, http_post, parse_json_body
from ..types import ChannelMessage, MessageDirection


class BlueBubblesChannel(WebhookChannel):
    """BlueBubbles server adapter (iMessage / SMS via macOS bridge).

    Inbound: BlueBubbles server POSTs ``new-message`` events.
    Outbound: ``POST {server}/api/v1/message/text?password=...``.
    """

    name = "bluebubbles"

    def __init__(
        self,
        server_url: str,
        password: str,
        *,
        http_client: tp.Any = None,
    ) -> None:
        """Bind the adapter to a BlueBubbles server.

        Args:
            server_url: Base URL of the BlueBubbles API (trailing ``/`` ok).
            password: Shared password used as a query-string credential.
            http_client: Optional injected HTTP callable for tests.
        """
        super().__init__()
        self.server_url = server_url.rstrip("/")
        self.password = password
        self._http = http_client

    def _parse_inbound(self, headers, body):
        """Convert a BlueBubbles ``new-message`` envelope into messages.

        Reads ``text`` / ``body`` as the content, ``chats[0].guid`` as the
        conversation id, and ``handle.address`` as the sender.
        """
        data = parse_json_body(body)
        if not data:
            return []
        msg_data = data.get("data", data)
        text = msg_data.get("text") or msg_data.get("body") or ""
        chat = msg_data.get("chats", [{}])[0] if isinstance(msg_data.get("chats"), list) else msg_data.get("chat", {})
        handle = msg_data.get("handle") or {}
        if not text:
            return []
        return [
            ChannelMessage(
                text=text,
                channel=self.name,
                channel_user_id=handle.get("address", ""),
                room_id=chat.get("guid", ""),
                platform_message_id=msg_data.get("guid", ""),
                direction=MessageDirection.INBOUND,
            )
        ]

    async def _send_outbound(self, message):
        """POST ``message`` to ``/api/v1/message/text`` using the private-api method."""
        url = f"{self.server_url}/api/v1/message/text?password={self.password}"
        body = {"chatGuid": message.room_id, "message": message.text, "method": "private-api"}
        http_post(url, json_body=body, http_client=self._http)

File: xerxes/src/channels/blueBubbles.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { postJson, providerUrl, type ChannelFetch } from './http.js'
import {
  arrayValue,
  outboundDestination,
  recordValue,
  requiredOption,
  stringValue,
  urlWithQuery,
  type RelayChannelTransport,
} from './relay.js'
import { createChannelMessage, MessageDirection, type ChannelMessage } from './types.js'
import { parseJsonBody, WebhookChannel, type WebhookHeaders } from './webhooks.js'

/** BlueBubbles webhook relay and HTTP sender; no server event socket is owned. */
export const BLUEBUBBLES_TRANSPORT: RelayChannelTransport = {
  inbound: 'webhook-relay',
  outbound: 'http-api',
  unsupported: ['BlueBubbles persistent event socket', 'attachment-only message delivery'],
}

export interface BlueBubblesChannelOptions {
  readonly fetchImplementation?: ChannelFetch
  readonly password: string
  readonly serverUrl: string
}

/**
 * Self-hosted BlueBubbles iMessage webhook relay and REST sender.
 *
 * Inbound events must be pushed by BlueBubbles to Xerxes. The adapter does
 * not establish the bridge's persistent event socket, and never logs URLs
 * that contain the server password.
 */
export class BlueBubblesChannel extends WebhookChannel {
  readonly name = 'bluebubbles'
  readonly transport = BLUEBUBBLES_TRANSPORT

  private readonly fetchImplementation: ChannelFetch | undefined
  private readonly password: string
  private readonly serverUrl: string

  constructor(options: BlueBubblesChannelOptions) {
    super()
    this.fetchImplementation = options.fetchImplementation
    this.password = requiredOption(options.password, 'BlueBubbles password')
    this.serverUrl = requiredOption(options.serverUrl, 'BlueBubbles serverUrl')
  }

  protected parseInbound(
    _headers: WebhookHeaders,
    body: Uint8Array,
  ): readonly ChannelMessage[] {
    const payload = parseJsonBody(body)
    const data = Object.keys(recordValue(payload.data)).length ? recordValue(payload.data) : payload
    const text = stringValue(data.text) || stringValue(data.body)
    if (!text) {
      return []
    }
    const chats = arrayValue(data.chats)
    const firstChat = chats.length ? recordValue(chats[0]) : recordValue(data.chat)
    const handle = recordValue(data.handle)
    return [createChannelMessage({
      channel: this.name,
      direction: MessageDirection.INBOUND,
      text,
      channelUserId: stringValue(handle.address),
      roomId: stringValue(firstChat.guid),
      platformMessageId: stringValue(data.guid),
      metadata: { is_from_me: data.isFromMe === true },
    })]
  }

  protected async sendOutbound(message: ChannelMessage): Promise<void> {
    const chatGuid = outboundDestination(message, 'BlueBubbles')
    const endpoint = urlWithQuery(
      providerUrl(this.serverUrl, 'api/v1/message/text'),
      'password',
      this.password,
    )
    await postJson(endpoint, {
      body: { chatGuid, message: message.text, method: 'private-api' },
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
    })
  }
}
